Add support for lambda to upload file

charleswong28 opened the pull request:
Fix [#1419](https://github.com/apollographql/apollo-server/issues/1419), [#1703](https://github.com/apollographql/apollo-server/issues/1703) file upload using lambda.

TODO:
* [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
* [x] Make sure all of the significant new logic is covered by tests
* [x] Rebase your changes on master so that they can be merged easily
* [x] Make sure all tests and linter rules pass

<!--**Pull Request Labels**

While not necessary, you can help organize our pull requests by labeling this issue when you open it.  To add a label automatically, simply [x] mark the appropriate box below:

- [x] feature
- [] blocking
- [ ] docs

To add a label not listed above, simply place `/label another-label-name` on a line by itself.
-->

Graphql and apollo-server are still quite new to me. It would be great if anyone here could point me to the right direction.

1. uploadsConfig is now passed to createHandler but I believe it's better to be passed to apollo server instead. Not sure how to access apollo server config in handler. 
2. Unit test was using apollo-server-integration-testsuite but I am not sure if other package is able to test with file upload or not.

charleswong28 commented:
Encountered a problem with uploading image in my project. 
```
Error: Part terminated early due to unexpected end of multipart data
 at ...node_modules/apollo-server-lambda/node_modules/@apollographql/apollo-upload-server/node_modules/dicer/lib/Dicer.js:65:36
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

Changing ```request.push(event.body);``` to ```request.push(Buffer.from(event.body, 'ascii'));``` fixed the problem. From my understanding, there is null character in event.body, the stream is closed earlier and the multipart data is ended unexpectedly. A new commit [(6f28f8b)](https://github.com/apollographql/apollo-server/pull/1739/commits/6f28f8b5e0c0482205d443bbffc779bf2252745f) is pushed.

charleswong28 commented:
According to [this](https://github.com/bbilger/jrestless-examples/blob/master/aws/gateway/aws-gateway-binary/README.md), `event.isBase64Encoded` will be `true` after adding `multipart/form-data` to Binary Media Types. In order to encode the stream correctly, I changed to
```
Buffer.from(event.body, event.isBase64Encoded ? 'base64' : 'ascii')
```
in this [commit](https://github.com/apollographql/apollo-server/pull/1739/commits/fbd9d664a3e92831ff69e125f75cf6f2dff9bdb9).

gorillatron commented:
So close! In need of this fix! :D Please

amwill04 commented:
@charleswong28 have been using this in development for the past week and havent had any issues so far FWIW

patrickr commented:
Any movement on reviewing this PR? my work progress is blocked by the issues this fixes

amwill04 commented:
Can someone advise if I can take over this PR? As far as I am aware it has passed all test and simply requires to be rebased with master - is this correct?

charleswong28 commented:
> Can someone advise if I can take over this PR? As far as I am aware it has passed all test and simply requires to be rebased with master - is this correct?
> 
> Im obviously not want to take any credit from @charleswong28 - I just simply need this fix ASAP

I am not sure uploadsConfig and unit test are the best way to do it. Hope there are some suggestions from the team. Anyway, I've squashed and rebased to master. 

My production branch also needs this fix ASAP. It's quite troublesome to use forked local branch currently when deploying this on lambda.

amwill04 commented:
@charleswong28 it is troublesome. MY work around was the following:

- update`pacakge.json` in fork to:

```
// Basically rewrite the dependencies
{
  "name": "apollo-server-lambda",
  "description": "Production-ready Node.js GraphQL server for AWS Lambda",
  "keywords": [
    "GraphQL",
    "Apollo",
    "Server",
    "Lambda",
    "Javascript"
  ],
  "author": {
    "name": "opensource@apollographql.com"
  },
  "bugs": {
    "url": "https://github.com/apollographql/apollo-server/issues"
  },
  "bundleDependencies": false,
  "dependencies": {
    "@apollographql/graphql-playground-html": "^1.6.0",
    "apollo-server-core": "^2.1.0",
    "apollo-server-env": "^2.0.3",
    "graphql-tools": "^3.0.4"
  },
  "deprecated": false,
  "description": "Production-ready Node.js GraphQL server for AWS Lambda",
  "devDependencies": {
    "apollo-server-integration-testsuite": "^2.1.0"
  },
  "engines": {
    "node": ">=6"
  },
  "gitHead": "86dd95e223460f8e1768141a8cefa58bcd409b03",
  "homepage": "https://github.com/apollographql/apollo-server#readme",
  "keywords": [
    "GraphQL",
    "Apollo",
    "Server",
    "Lambda",
    "Javascript"
  ],
  "license": "MIT",
  "main": "dist/index.js",
  "name": "apollo-server-lambda",
  "peerDependencies": {
    "graphql": "^0.12.0 || ^0.13.0 || ^14.0.0"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/apollographql/apollo-server/tree/master/packages/apollo-server-lambda"
  },
  "scripts": {
    "clean": "rm -rf dist",
    "compile": "tsc",
    "prepare": "npm run clean && npm run compile"
  },
  "types": "dist/index.d.ts",
  "version": "2.1.0"
}

```

- `yarn prepare`
- `tar.gz` `apollo-server-lambda` and move to my project folder.
- edit project `pacakge.json`

```
"dependencies": {
    "apollo-server-lambda": "file:./apollo-server-lambda.tar.gz",
}
```

ariariel commented:
Any updates on this PR being merged?

k-komarov commented:
Any updates?

kaamoss commented:
Any updates on this? Like everyone else here I have a few projects that could benefit a lot from this.

JorgeCeja commented:
@amwill04 or charleswong28 I can't get the fork working on my end. Do you guys have a simple example of it working? I keep getting empty objects back in the upload mutatiton. The issue is not in the front end because it works if I use the other apollo-server which support file uploads. 

Thanks!

livingmine commented:
@abernix there is an update on this PR to implement things stated on [your comment](https://github.com/apollographql/apollo-server/pull/1739#issuecomment-459452145) in order to get this into a release. Do you have any comment about the update so that we can get this into release as soon as possible? Thanks!

bluedusk commented:
@charleswong28  Hi mate is this PR still in progress ? I got this issue too and wondering can I try to use your branch?  Is this working with apollo-client-upload ? 
Thanks

bluedusk commented on packages/apollo-server-lambda/src/lambdaApollo.ts:
```diff
@@ -38,13 +38,20 @@ export function graphqlLambda(
+    let query: any = event.queryStringParameters;
+    if (event.httpMethod === 'POST' && event.body) {
+      if (typeof event.body === 'string') {
+        query = JSON.parse(event.body);
```
This will thow an error when I try use apollo-client-upload. Since the content-type is multipart, event.body is a string but not a json.  Not sure how can this be solved ? Or will this support multipart ?

charleswong28 commented:
@bluedusk 
I guess it's the setting in AWS API gateway. Did you set multipart/form-data to Binary Media Types in API Gateway?

If you are using multipart/form-data, it works with Lambda Proxy Integration and adding multipart/form-data to Binary Media Types. You will then see event.isBase64Encoded to be true. We parse the event.body using base64 in ApolloServer.ts fileUploadProcess function.
```
Buffer.from(event.body, event.isBase64Encoded ? 'base64' : 'ascii')
```

You may find [this](https://medium.com/@lakshmanLD/lambda-proxy-vs-lambda-integration-in-aws-api-gateway-3a9397af0e6d) and [this](https://github.com/bbilger/jrestless-examples/blob/master/aws/gateway/aws-gateway-binary/README.md) helpful. 

If you are using serverless, we can use [serverless-apigw-binary](https://github.com/maciejtreder/serverless-apigw-binary) and [serverless-content-encoding](https://github.com/dong-dohai/serverless-content-encoding) plugins.
```
custom:
  apigwBinary:
    types:
      - 'image/gif'
      - 'application/octet-stream'
      - 'multipart/form-data'
  contentEncoding:
    # 30KB
    minimumCompressionSize: 30720
```

I hope this will help. We've been using this for our production graphql so I guess you can try. event.isBase64Encoded should be true and it will be parsed successfully after setting multipart/form-data to Binary Media Type in API Gateway.

jalie commented:
@charleswong28 I am using your apollo-server (serverless-offline) with your recommended serverless-apigw-binary and serverless-content-encoding package + custom config.

but i am still getting the JSON error:
`
[offline] SyntaxError: Unexpected number in JSON at position 1
  at JSON.parse (<anonymous>)
  at graphqlHandler (/**/node_modules/apollo-server-lambda/dist/lambdaApollo.js:24:24)
`

Might there be something else needed?

bluedusk commented on packages/apollo-server-lambda/src/ApolloServer.ts:
```diff
@@ -145,33 +152,88 @@ export class ApolloServer extends ApolloServerBase {
+      contentType &&
+      contentType.startsWith('multipart/form-data')
+    ) {
+      const request = new stream.Readable() as any;
```
I have an error here. got it work by changing to new Stream.Readable().  
typo ?

bluedusk replied:
@charleswong28 Would like to confirm with you, any comments on this ?

charleswong28 commented:
@jalie @bluedusk I think we need to create an example repository for this but I am not available today. Will try it this weekend.

bluedusk commented:
> @jalie @bluedusk I think we need to create an example repository for this but I am not available today. Will try it this weekend.

Thanks mate, I could use some time to build a example repo. Any idea how could I install your version of apollo-server-lambda with npm rather than copy your code to my project ?

charleswong28 commented:
> > @jalie @bluedusk I think we need to create an example repository for this but I am not available today. Will try it this weekend.
> 
> Thanks mate @charleswong28 , I could use some time to build a example repo. Any idea how could I install your version of apollo-server-lambda with npm rather than copy your code to my project ?

That would be very helpful. Sorry that I still couldn't squeeze time out for this yet.
The comment from amwill04 would also work by compressing the repository but the package.json has typo when copying and pasting. You will see when you copy it. The idea is basically build it and then use `"apollo-server-lambda": "file:./apollo-server-lambda.tar.gz",` to refer to it.

Another way would be cloning it next to your example repository and use `"apollo-server-lambda": "file:./apollo-server/packages/apollo-server-lambda",` to refer to it. This will be easier when you change code inside apollo-server-lambda repository.

bluedusk commented:
Create a demo with your PR Code:
A apollo server fileupload demo using apollo-server-lambda:
https://github.com/bluedusk/apollo-server-lambda-fileupload

charleswong28 commented:
> Create a demo with your PR Code:
> A apollo server fileupload demo using apollo-server-lambda:
> https://github.com/bluedusk/apollo-server-lambda-fileupload

It shows internal server error due to service timeout and "src" is missing in your postman.json. I've created a pull request.
https://github.com/bluedusk/apollo-server-lambda-fileupload/pull/1

In my branch, I am able to see the 
```
[ Promise {
    { filename: 'package.json',
    mimetype: 'application/json',
    encoding: '7bit',
    createReadStream: [Function: createReadStream] } } ]
```
in uploadFiles resolver files param. 

Hope this help!

JorgeCeja commented:
Any updates on when this PR could be merged to master?

bluedusk commented:
Hey guys, Anyone from the team following this PR ?

eliliam commented:
Is this PR going to get merged or has it fallen stale?

intellix commented on packages/apollo-server-lambda/package-lock.json:
```diff
@@ -0,0 +1,1927 @@
+  "requires": true,
+  "dependencies": {
+    "@apollographql/apollo-upload-server": {
+      "version": "5.0.3",
```
this was replaced with https://github.com/jaydenseric/graphql-upload I suppose

eliliam commented:
I'm all down for using graphql-upload for now, but don't you think it would be better to merge this into master for sake of continuity between implementations?

abernix commented on packages/apollo-server-lambda/src/ApolloServer.ts:
```diff
@@ -173,30 +180,86 @@ export class ApolloServer extends ApolloServerBase {
+            // to `await` the `promiseWillStart` which we kicked off at the top of
+            // this method to ensure that it runs to completion (which is part of
+            // its contract) prior to processing the request.
+            await promiseWillStart;
```
The `promiseWillStart` must be awaited prior to entering the `fileUploadProcess`.  We can't wait until after processing the file upload to fire this callback.

abernix commented on packages/apollo-server-lambda/src/ApolloServer.ts:
```diff
@@ -173,30 +180,86 @@ export class ApolloServer extends ApolloServerBase {
-      })(event, context, callbackFilter);
+      const response = new Stream.Writable();
+
+      fileUploadProcess(event, response, this.uploadsConfig || {})
```
The code here would seem to work, but this is a confusing arrangement of logic.  The `fileUploadProcess` shouldn't be the thing that kicks off the chain, it should be an optional chain within the path of request handling.

To put it another way, let's maintain that `graphqlLambda` is the thing that gets kicked off, and the callback passed to `graphqlLambda` be the part of the code that brings the upload processing`, rather than having `graphqlLambda` triggered within `fileUploadProcess` — that's not an intuitive place for it.

abernix commented on packages/apollo-server-lambda/src/ApolloServer.ts:
```diff
@@ -173,30 +180,86 @@ export class ApolloServer extends ApolloServerBase {
   }
 }
+
+const fileUploadProcess = (
```
This name is too close to `processFileUploads`.  This might be better named as a `processRequest`, but that's not even quite right.  I realize that this code left some to be desired before, but I'm afraid that this makes it quite difficult to understand.  There are several other frameworks to use as a basis, but this is pretty far from all of them right now.

jasonlimantoro commented:
Any updates for this PR?
Just for anyone who stumbles around this issue, I have tested that this PR works with `apollo-upload-client` package

mrfoster commented:
I would love to see this get finished off, it's so close yet such slow progress..

I have had a real headache getting this working:

1) Ran `npm i` in root folder followed by `npm link` in the apollo-server-lambda folder. This resulted in a weird error trying to rename a non existent folder (.staging). After a bit of searching,  I overcame it by removing the package.lock file. This time node ran out of memory and after increasing the limit to 8000mb, I discovered that it was just endlessly looping around dependencies.  I am using the latest LTS node and npm and have no idea how to overcome it so gave up.

2) I copied the folder into my repo and got it working after switching a couple of imports to `import * as`.

3) I then discovered that CORS was broken but, after a bit of digging around, I applied the following commit and CORS was fixed.  https://github.com/apollographql/apollo-server/commit/7e78a2353b9e855d85a7998b06e8f9d4cb1db54d#diff-16d4d9e1275e5b29399d54c22f79316f

4) I re-enabled the upload config and got the following error `This implementation of ApolloServer does not support file uploads because the environment cannot accept multi-part forms`. This was a little easier to track down and was fixed by adding 
```
  supportsUploads(): boolean {
    return true;
  }
```

5) It now appears to work but the files seem to be corrupt. I have checked that multipart/form-data has been added to BinaryMediaTypes so I'm a little lost but will hopefully track it down today and post back.

Hopefully this will help someone if they are desperate enough to use it now.

mrfoster commented:
Turned out that I was handling base64 encoded files elsewhere in my code.  I can confirm that it is now working by applying steps 3 and 4 from https://github.com/apollographql/apollo-server/pull/1739#issuecomment-518531156

zakirbaytar commented:
Any updates on this PR?

mikemeerschaert commented:
What's left on this preventing a merge? I need to add a feature to handle file uploads in my project very soon and I'd like to do it the officially supported way so it doesn't have to be changed later. Are there specific action items that need to be addressed before this gets merged?

aymericbouzy commented:
@charleswong28 I see the last commit was on Feb 14th 2019 : are you still planning to work on this PR? is it ok if I open a new PR with commits to address the requested changes by @abernix?

jurienhamaker commented:
Would love to take over this PR if @charleswong28 is not planning on fixing it further.

charleswong28 commented:
> Would love to take over this PR if @charleswong28 is not planning on fixing it further.

@jurienhamaker  @aymericbouzy Sorry for not having enough time to work on this. It would be great if you guys could help.

aymericbouzy commented:
@charleswong28 no worries, we totally understand 😄 
@jurienhamaker you seem highly motivated, please proceed 😊 I'm new to the repo, but if you want any help, please ask!

k00k commented:
I've created a new PR #3676 that uses some of @charleswong28's work from this PR and attempts to incorporate the comments by @abernix on this PR.  All checks have passed, so hoping to get it merged pretty quickly as we actually need to use it :) 

Feel free to chime in over there.

abernix commented:
Superseded by https://github.com/apollographql/apollo-server/pull/3926, which has merged into `release-2.13.0`.  An _alpha_ release has been published on the `alpha` npm tag which includes #3926, and can be installed with `npm install apollo-server-lambda@2.13.0-alpha.0`.  Please try it out and report any issues on a new issue which references #3926.
